### DISCUSSION TITLE
如何正确使用useModel

### DISCUSSION THREAD

**Discussion by 262146150** [umijs/umi#11219] [General]
umi max 项目中，首先使用了自定义layout，并且创建了两个model，在组件中useModel没有任何问题，但是在login.tsx组件中，没有使用 useModel，也执行了所有 model里面的 useEffect 方法，我想问的是，是不是只要有models，在任何一个组件中都会执行？如果不是，应该如何指定在某些组件中，不需要执行 models中的方法？
`export default function Page() {

    const [user, setUser] = useState<API.userProfile>({});

    const updateInfo = useCallback((userInfo: API.userProfile) => {
        console.log(userInfo)
        setUser({ ...userInfo })
    }, [user])

    const getUserInfo = async () => {
        try {
            const { data } = await queryUserInfo({});
            setUser(data)
        } catch (error: any) {

        } finally {

        }
    };

    useEffect(() => {
        if (history.location.pathname !== '/login' && history.location.pathname !== '/auth') {
            getUserInfo()
        }
    }, []);

    return { user, updateInfo };
};`

比如这个 userModels，任意组件中都会执行，useEffect，因此为了避免，我增加了根据 pathname判断哪些路由需要执行，哪些不需要，但是觉得不太合适，是否有更简单的办法？

**Comment by yourSwiftie** [comment]
同问

**Comment by stormslowly** [comment]
useModels  返回一个更新 state 函数 (有副作用的)，这个函数在项目合适的地方调用，更新对应的数据。

比如 获取用户信息，可以放在 layout 里面，其他组件通过 useModels 获取用户信息即可。

（不知道我理解的对不对  ：p

**Reply by 262146150** [reply]
你说的没错，但是我的问题是，我没在 login 组件中使用models，但是当访问login页面的时候，也调用了models里面的useEffect方法了，其实并想这样

**Comment by xliez** [comment]
+1，同问。

目前在没有调用useModel的页面下也会去执行useModel。

我遇到的情况：

在models中写了几个获取一些状态的异步接口，结果在登录页（没有调用useModel）就会去执行请求，然后自然接口就报错啦。

**Reply by xliez** [reply]
手动 at 下大佬 @fz6m

**Reply by 262146150** [reply]
我们应该是同一种问题

**Reply by fz6m** [reply]
useModel 本质上是找了一个持久性的、全局的 context 来存储所有的 hooks ：

 - 持久性的：本来 hooks 是不能跨组件共享东西的，正因为持久的存在了，所以你才可以跨组件的使用，这就意味着所有的 hooks 一直存在，使得 effect 会默认就执行。

 - 全局的：目前所有的 model 都放在全局，没有局部作用域这个概念，如果你希望局部的 model ，可以尝试用 [hox](https://github.com/umijs/hox) 这个解决方案，他和 umi 里的 model 类似，但作用域是局部的，使用体验类似，但这个方案是有一些 edge case 的，如果需要局部的 effect ，应该在某个局部的 context 去运行，另外，更推荐使用现代的全局数据存储方案 [valtio](https://umijs.org/docs/max/valtio) / [jotai](https://github.com/pmndrs/jotai) 来管理全局数据，用 [react-query](https://umijs.org/docs/max/react-query) 来管理请求状态，从而很少会用到 model 这种全局的共享场景。

**Reply by 262146150** [reply]
大佬解释的很明白，懂了懂了，是否能在 Umijs 的文档中，把一些“推荐”的方案加个tag，这样我们会更有方向，省的大佬在各种回复中重复的说“推荐”

**Reply by samous0726** [reply]
我的UserModel为何会执行多次？

**Reply by Bclound** [reply]
我现在也遇到这个问题了，你解决了吗？

**Reply by 262146150** [reply]
不用解决，上面解释了原因，我暂时使用path做了个简单的判断

**Reply by Bclound** [reply]
我现在倒是没有用到useEffect，是models多次渲染，导致我state设置的值又被初始化了。。
![image](https://github.com/umijs/umi/assets/57136514/aee1b7c9-fac5-4c53-bc1d-aa6ebfb3f09a)

**Comment by purple-force** [comment]
> umi max 项目中，首先使用了自定义layout，并且创建了两个model，在组件中useModel没有任何问题，但是在login.tsx组件中，没有使用 useModel，也执行了所有 model里面的 useEffect 方法，我想问的是，是不是只要有models，在任何一个组件中都会执行？如果不是，应该如何指定在某些组件中，不需要执行 models中的方法？ `export default function Page() {
> 
> ```
> const [user, setUser] = useState<API.userProfile>({});
> 
> const updateInfo = useCallback((userInfo: API.userProfile) => {
>     console.log(userInfo)
>     setUser({ ...userInfo })
> }, [user])
> 
> const getUserInfo = async () => {
>     try {
>         const { data } = await queryUserInfo({});
>         setUser(data)
>     } catch (error: any) {
> 
>     } finally {
> 
>     }
> };
> 
> useEffect(() => {
>     if (history.location.pathname !== '/login' && history.location.pathname !== '/auth') {
>         getUserInfo()
>     }
> }, []);
> 
> return { user, updateInfo };
> ```
> 
> };`
> 
> 比如这个 userModels，任意组件中都会执行，useEffect，因此为了避免，我增加了根据 pathname判断哪些路由需要执行，哪些不需要，但是觉得不太合适，是否有更简单的办法？

在 umi 里的 model 都是全局性质的，也就是说这里面的东西和副作用在应用启动一开始就会执行了。

可以看看官方的回复：https://github.com/umijs/umi/discussions/11824
